src/gpu_sysman.c

```diff
@@ -0,0 +1,1939 @@
+    free(gpus[i].pci_bdf);
+    gpus[i].pci_bdf = NULL;
+    free(gpus[i].dev_file);
+    gpus[i].dev_file = NULL;
+  }
+#undef FREE_GPU_SAMPLING_ARRAYS
+#undef FREE_GPU_ARRAY
+  free(gpus);
+  gpus = NULL;
+  return RET_OK;
+}
+
+/* show plugin GPU metrics config options, return RET_OK
+ * if at least some metric is enabled, otherwise error code
+ */
+static int gpu_config_check(void) {
+  if (config.output == OUTPUT_UNSET) {
+    config.output = OUTPUT_BOTH;
+  }
+  assert(config.output < STATIC_ARRAY_SIZE(metrics_output));
+
+  if (config.gpuinfo) {
+    INFO("Sysman '" KEY_SAMPLES "': %d", config.samples);
+    INFO(KEY_METRICS_OUTPUT ": %s", metrics_output[config.output]);
+    INFO("Disabled metrics:");
+  }
+  struct {
+    const char *name;
+    bool value;
+  } options[] = {{KEY_DISABLE_ENGINE, config.disabled.engine},
+                 {KEY_DISABLE_ENGINE_SINGLE, config.disabled.engine_single},
+                 {KEY_DISABLE_FREQ, config.disabled.freq},
+                 {KEY_DISABLE_MEM, config.disabled.mem},
+                 {KEY_DISABLE_MEMBW, config.disabled.membw},
+                 {KEY_DISABLE_POWER, config.disabled.power},
+                 {KEY_DISABLE_RAS, config.disabled.ras},
+                 {KEY_DISABLE_RAS_SEPARATE, config.disabled.ras_separate},
+                 {KEY_DISABLE_TEMP, config.disabled.temp},
+                 {KEY_DISABLE_THROTTLE, config.disabled.throttle}};
+  unsigned int i, disabled = 0;
+  for (i = 0; i < STATIC_ARRAY_SIZE(options); i++) {
+    if (options[i].value) {
+      if (config.gpuinfo) {
+        INFO("- %s", options[i].name);
+      }
+      disabled++;
+    }
+  }
+  if (disabled >= STATIC_ARRAY_SIZE(options)) {
+    ERROR(PLUGIN_NAME ": all metrics disabled");
+    return RET_NO_METRICS;
+  }
+  if (config.gpuinfo) {
+    if (disabled) {
+      INFO("=> %d disabled metrics", disabled);
+    } else {
+      INFO("- no disabled metrics");
+    }
+  }
+  return RET_OK;
+}
+
+/* Set GPU specific flags to initial global configuration values
+ * for each GPU.  Allocations of metrics arrays are done when metrics
+ * are queried for the first time (not here), and re-allocated if
+ * number of types for given metric changes.
+ *
+ * Return RET_OK if config is OK, (negative) error value otherwise
+ */
+static int gpu_config_init(unsigned int count) {
+  if (!config.samples) {
+    config.samples = 1;
+  }
+  if (gpu_config_check()) {
+    gpu_config_free();
+    return RET_NO_METRICS;
+  }
+  unsigned int i;
+  for (i = 0; i < count; i++) {
+    gpus[i].disabled = config.disabled;
+    gpus[i].counter = 0;
+  }
+  gpu_count = count;
+  return RET_OK;
+}
+
+/* log given UUID (without dashes):
+ * https://en.wikipedia.org/wiki/Universally_unique_identifier
+ */
+static void log_uuid(const char *prefix, const uint8_t *byte, int len) {
+  int offset = strlen(prefix);
+  char buf[offset + 2 * len + 1];
+  strcpy(buf, prefix);
+  while (len-- > 0) {
+    sprintf(buf + offset, "%02x", *byte++);
```

Comment:
https://github.com/collectd/collectd/blob/main/src/utils/common/common.c#L93

Reply:
IMHO "ssnprintf" wrapper is pointless, it just adds extra overhead & confusion compared to using just snprintf() directly.

As to using snprintf(), "%02x" on a byte does have a fixed size, so it's already safe.  Giving "sizeof(buf) - offset" as size for snprintf() is also somewhat error prone for future modifications.  I would prefer keeping that as strcpy() so that it gets highlighted when people review code.

(Best option would be for collectd having helper function for converting UUIDs to strings.)